### PR TITLE
Feature: Added initial value to non-treestated input pins (addresses issue #1477)

### PR DESCRIPTION
**Pull request by BFH-ktt1**

### CODE REVIEW

**Comment by MarcinOrlowski** [comment]
Please also add CHANGELOG entry

File: src/main/resources/resources/logisim/strings/std/std.properties [hunk 1]
```diff
@@ -858,6 +858,7 @@ pinPullAttr = Pull Behavior
 pinPullDownOption = Pull Down
 pinPullNoneOption = Unchanged
 pinPullUpOption = Pull Up
+pinInitialValue = Reset value:
```
Comment (MarcinOrlowski) [inline src/main/resources/resources/logisim/strings/std/std.properties:861]: The key is `initialValue` but the value is `Reset value`. Can we have this more consistent - would be less confusing for translators.

File: src/main/java/com/cburch/logisim/std/wiring/PinAttributes.java [hunk 2]
```diff
@@ -41,27 +42,36 @@ public boolean isHidden() {
...
+    myAttributes = new ArrayList<Attribute<?>>();
+    myAttributes.add(StdAttr.FACING);
+    myAttributes.add(Pin.ATTR_TYPE);
+    myAttributes.add(StdAttr.WIDTH);
+    myAttributes.add(Pin.ATTR_TRISTATE);
+    myAttributes.add(Pin.ATTR_PULL);
+    myAttributes.add(StdAttr.LABEL);
+    myAttributes.add(StdAttr.LABEL_FONT);
+    myAttributes.add(RadixOption.ATTRIBUTE);
+    if ((type == EndData.INPUT_ONLY) && !threeState) myAttributes.add(Pin.ATTR_INITIAL);
```
Comment (MarcinOrlowski) [inline src/main/java/com/cburch/logisim/std/wiring/PinAttributes.java:67]: Can we have this line as 

```java
if ((type == EndData.INPUT_ONLY) && !threeState) {
    myAttributes.add(Pin.ATTR_INITIAL);
}
```

for better readability?

File: src/main/java/com/cburch/logisim/std/wiring/Pin.java [hunk 3]
```diff
@@ -673,17 +673,18 @@ public Object clone() {
...
+    final var width = attrs.width;
+    var ret = (PinState) state.getData();
     if (ret == null) {
-      Value val = attrs.threeState ? Value.UNKNOWN : Value.FALSE;
-      if (width.getWidth() > 1) {
-        Value[] arr = new Value[width.getWidth()];
-        java.util.Arrays.fill(arr, val);
-        val = Value.create(arr);
+      var initialValue = attrs.getValue(ATTR_INITIAL);
+      Value newValue;
```
Comment (MarcinOrlowski) [inline src/main/java/com/cburch/logisim/std/wiring/Pin.java:681]: Can we have this block as

```java
      final var newValue = (attrs.threeState) 
            ? Value.createUnknown(width)
            : Value.createKnown(width.getWidth(), initialValue);
```

File: src/main/java/com/cburch/logisim/std/wiring/PinAttributes.java [hunk 4]
```diff
@@ -41,27 +42,36 @@ public boolean isHidden() {
...
-          RadixOption.ATTRIBUTE,
-          PROBEAPPEARANCE,
-          ATTR_DUMMY);
+  private List<Attribute<?>> myAttributes;
 
   BitWidth width = BitWidth.ONE;
   boolean threeState = false; // true;
   int type = EndData.INPUT_ONLY;
   Object pull = Pin.PULL_NONE;
+  Long initialValue = 0L;
```
Comment (MarcinOrlowski) [inline src/main/java/com/cburch/logisim/std/wiring/PinAttributes.java:51]: Do we need `Long` here and not `long` primitive? If so `var` should be inferred from value so there would not be a need for explicit type declaration.

**Review by MarcinOrlowski — CHANGES_REQUESTED** [review]

File: src/main/resources/resources/logisim/strings/std/std.properties (hunk 1 above)
Comment (BFH-ktt1) [reply src/main/resources/resources/logisim/strings/std/std.properties:861]: Won´t fix, Initial or Reset is for me the same

File: src/main/java/com/cburch/logisim/std/wiring/PinAttributes.java (hunk 2 above)
Comment (BFH-ktt1) [reply src/main/java/com/cburch/logisim/std/wiring/PinAttributes.java:67]: done

File: src/main/java/com/cburch/logisim/std/wiring/PinAttributes.java (hunk 4 above)
Comment (BFH-ktt1) [reply src/main/java/com/cburch/logisim/std/wiring/PinAttributes.java:51]: done

File: src/main/java/com/cburch/logisim/std/wiring/Pin.java (hunk 3 above)
Comment (BFH-ktt1) [reply src/main/java/com/cburch/logisim/std/wiring/Pin.java:681]: done

File: src/main/resources/resources/logisim/strings/std/std.properties (hunk 1 above)
Comment (maehne) [reply src/main/resources/resources/logisim/strings/std/std.properties:861]: I concur with @MarcinOrlowski in this matter. I think "initial value" is more appropriate. Also, there's a semantical difference in the context of discrete event simulators: Initial value signifies the initial value of the signal at start of simulation after end of elaboration. The reset value is the value imposed if a corresponding reset signal becomes active and imposes the change of the signal value.

File: src/main/resources/resources/logisim/strings/std/std_de.properties [hunk 5]
```diff
@@ -858,6 +858,7 @@ pinPullAttr = Pull-Verhalten
 pinPullDownOption = Pull-Down
 pinPullNoneOption = Unverändert
 pinPullUpOption = Pull-Up
+pinInitialValue = Reset Wert:
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/std/std_de.properties:861]: Startwert:

File: src/main/resources/resources/logisim/strings/std/std_fr.properties [hunk 6]
```diff
@@ -858,6 +858,7 @@ pinPullAttr = Comportement
 pinPullDownOption = Rappel (pull-down)
 pinPullNoneOption = Inchangé
 pinPullUpOption = Tirage (pull-up)
+# pinInitialValue =
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/std/std_fr.properties:861]: Valeur initiale :

File: src/main/java/com/cburch/logisim/std/wiring/PinAttributes.java (hunk 4 above)
Comment (maehne) [reply src/main/java/com/cburch/logisim/std/wiring/PinAttributes.java:51]: @BFH-ktt1: At least on GitHub, you don't seem to have taken into account @MarcinOrlowski's suggestion.

**Review by maehne — CHANGES_REQUESTED** [review]
Looks mostly good to me except for the localisation strings, where I concur with @MarcinOrlowski.

File: src/main/resources/resources/logisim/strings/std/std_de.properties (hunk 5 above)
Comment (BFH-ktt1) [reply src/main/resources/resources/logisim/strings/std/std_de.properties:861]: done

File: src/main/resources/resources/logisim/strings/std/std_fr.properties (hunk 6 above)
Comment (BFH-ktt1) [reply src/main/resources/resources/logisim/strings/std/std_fr.properties:861]: done

File: src/main/resources/resources/logisim/strings/std/std_cn.properties [hunk 7]
```diff
@@ -858,6 +858,7 @@ pinPullAttr = 拉动行为
 pinPullDownOption = 下拉
 pinPullNoneOption = 不变
 pinPullUpOption = 向上拉起
+# pinResetValue =
```
Comment (MarcinOrlowski) [inline src/main/resources/resources/logisim/strings/std/std_cn.properties:861]: > `# pinResetValue =`

Don't we rather use `==>` prefixed keys (multiple places)?

    # ==> pinResetValue =

**Review by MarcinOrlowski — CHANGES_REQUESTED** [review]

File: src/main/resources/resources/logisim/strings/std/std_cn.properties (hunk 7 above)
Comment (BFH-ktt1) [reply src/main/resources/resources/logisim/strings/std/std_cn.properties:861]: Done

Comment (BFH-ktt1) [reply src/main/resources/resources/logisim/strings/std/std_cn.properties:861]: Sorry, but there is now a small problem that my pushes to my fork are now related to another repro. Have to sort that out.

Comment (BFH-ktt1) [reply src/main/resources/resources/logisim/strings/std/std_cn.properties]: Now it is fixed

File: src/main/resources/resources/logisim/strings/std/std.properties (hunk 1 above)
Comment (MarcinOrlowski) [reply src/main/resources/resources/logisim/strings/std/std.properties]: My point is that by work meaning `initial value` and `reset value` are not equivalent. You can start with `initial value` and then reset to `reset value` and these two do not need to be the same. Also for me `reset value` does not automatically implies that this is also the `initial value`. Can we therefore meet in-between and set value to `Initial/reset value`? That would **explicitly** cover both cases and left no room for misinterpretation. 

@davidhutchens @maehne @BFH-ktt1

Comment (BFH-ktt1) [reply src/main/resources/resources/logisim/strings/std/std.properties]: @MarcinOrlowski : I agree with your discussion, would, however, not take `initial value` as this is a programming term. Real hardware does not have an `initial value`, but is set to a defined state by a reset. Hence, therefore I would opt to stay with `reset value`. What do you think?

Comment (MarcinOrlowski) [reply src/main/resources/resources/logisim/strings/std/std.properties]: @BFH-ktt1 I am buying the "real hardware" argument as justification for that wording. Let's move on. Thanks.

**Review by MarcinOrlowski — DISMISSED** [review]

**Review by MarcinOrlowski — DISMISSED** [review]

**Comment by BFH-ktt1** [comment]
@maehne : What is still blocking this PR?

**Pull request merged**